src/sqlfluff/utils/reflow/elements.py

```diff
@@ -0,0 +1,383 @@
+# logs to appear in the context of rules. Hence it's a subset
+# of the rules logger.
+reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
+
+
+@dataclass(frozen=True)
+class ReflowElement:
+    """Base reflow element class."""
+
+    segments: Sequence[RawSegment]
+
+    @staticmethod
+    def _class_types(segments: Sequence[RawSegment]) -> Set[str]:
+        return set(chain.from_iterable(seg.class_types for seg in segments))
+
+    @property
+    def class_types(self):
+        """The set of contained class types.
+
+        Parallel to BaseSegment.class_types
+        """
+        return self._class_types(self.segments)
+
+
+@dataclass(frozen=True)
+class ReflowBlock(ReflowElement):
+    """Class for keeping track of elements to reflow.
+
+    It holds segments to reflow and also exposes configuration
+    around how they are expected to reflow around others.
+
+    The attributes exposed are designed to be "post configuration"
+    i.e. they should reflect configuration appropriately.
+
+    NOTE: These are the smallest unit of "work" within
+    the reflow methods, and may contain meta segments.
+    """
+
+    # Options for spacing rules are:
+    # - single:         the default (one single space)
+    # - touch:          no whitespace
+    # - any:            don't enforce any spacing rules
+    spacing_before: str
+    spacing_after: str
+    # The depth info is used in determining where to put line breaks.
+    depth_info: DepthInfo
+    # This stores relevant configs for segments in the stack.
+    stack_spacing_configs: Dict[int, str]
+
+    @classmethod
+    def from_config(cls, segments, config: ReflowConfig, depth_info: DepthInfo):
+        """Extendable constructor which accepts config."""
+        block_config = config.get_block_config(cls._class_types(segments))
+        # Populate any spacing_within config.
+        # TODO: This needs decent unit tests - not just what happens in rules.
+        stack_spacing_configs = {}
+        for hash, class_types in zip(
+            depth_info.stack_hashes, depth_info.stack_class_types
+        ):
+            spacing_within = config.get_block_config(class_types).get(
+                "spacing_within", None
+            )
+            if spacing_within:
+                stack_spacing_configs[hash] = spacing_within
+        return cls(
+            segments=segments,
+            spacing_before=block_config.get("spacing_before", "single"),
+            spacing_after=block_config.get("spacing_after", "single"),
+            depth_info=depth_info,
+            stack_spacing_configs=stack_spacing_configs,
+        )
+
+
+@dataclass(frozen=True)
+class ReflowPoint(ReflowElement):
+    """Class for keeping track of editable elements in reflow.
+
+    It holds segments which can be changed during a reflow operation
+    such as whitespace and newlines.
+
+    It holds no configuration and is influenced by the blocks either
+    side.
+    """
+
+    def respace_point(
```

Comment:
Are there automated tests for this function? Would love to have some fairly "standalone" tests, as I've often found such tests useful with other lower-level components of the project.

Related: This function is pretty long. Maybe break it into smaller functions? (Seems good _eventually_, but it may make sense to wait until things have settled down more.)

Reply:
Yeah - it should be split apart, and those parts should have some more sensible low level tests. I'm going to think about how to approach this. 🤔

Reply:
I've added some explicit test cases for `ReflowPoint.respace_point()` and `ReflowSequence.respace()` (see `respace_test.py`), and I've split out a lot of the logic in the bigger functions into smaller methods, however: getting sensible test harnesses around those smaller functions feels like quite a big job, as each one relies on quite a lot of config under the hood.

See what I've done and whether you think it's clear enough?